File: frontend/src/models/project.ts
import { TaskRead } from "./task";
import { StatusRead } from "./status";
import { AssociationType } from "../enums";
import { ProjectAssociationDetail } from "./projectassociation";

export interface ProjectDetails extends ProjectCreate {
    id: string;
    tasks: TaskRead[];
    statuses: StatusRead[];
    workingTaskID?: string;
    associations: ProjectAssociationDetail[];
}

export interface ProjectCreate {
    name: string;
    abbreviation: string;
    description: string;
}

export interface ProjectGrid {
    id: string;
    name: string;
    abbreviation: string;
    associationType: AssociationType,
    lastUpdated: number
}

export interface ProjectTaskOrder {
    id: string,
    tasks: TaskOrder[],
}

export interface TaskOrder {
    id: string,
    statusId: string,
    ordinal: number
}

export const cloneProject = (oldProject: ProjectDetails) => {
    return {
        id: oldProject.id,
        name: oldProject.name,
        abbreviation: oldProject.abbreviation,
        description: oldProject.description,
        tasks: [...oldProject.tasks],
        statuses: [...oldProject.statuses],
        associations: [...oldProject.associations]
    }
}

File: pkg/api/data/project.go
package data

import (
	"github.com/volatiletech/null/v8"
)

type ProjectGrid struct {
	ID              string `json:"id"`
	AssociationType string `json:"associationType"`
	Name            string `json:"name"`
	Abbreviation    string `json:"abbreviation"`
	Description     string `json:"description"`
	LastUpdated     int64  `json:"lastUpdated"`
}

type ProjectDetail struct {
	ID            string                     `json:"id"`
	Name          string                     `json:"name"`
	Abbreviation  string                     `json:"abbreviation"`
	Description   string                     `json:"description"`
	Statuses      []StatusRead               `json:"statuses"`
	Tasks         []TaskGrid                 `json:"tasks"`
	WorkingTaskID null.String                `json:"workingTaskID"`
	Associations  []ProjectAssociationDetail `json:"associations"`
}

type ProjectCreate struct {
	Name         string `json:"name" binding:"required"`
	Abbreviation string `json:"abbreviation" binding:"required"`
	Description  string `json:"description" binding:"required"`
}

type ProjectUpdate struct {
	ID           string `json:"id" binding:"required"`
	Name         string `json:"name" binding:"required"`
	Abbreviation string `json:"identifier" binding:"required"`
	Description  string `json:"description" binding:"required"`
	Active       bool   `json:"active" binding:"required"`
}

type ProjectArchive struct {
	ProjectID string `json:"projectID" binding:"required"`
}

type ProjectTaskOrder struct {
	ID    string      `json:"id" binding:"required"`
	Tasks []TaskOrder `json:"tasks" binding:"required"`
}

type TaskOrder struct {
	ID       string `json:"id" binding:"required"`
	StatusID string `json:"statusId" binding:"required"`
	Ordinal  int    `json:"ordinal" binding:"required"`
}
